yt/fields/fluid_fields.py

```diff
@@ -195,12 +195,29 @@ def grad_func(axi, ax):
         slice_3dl = slice_3d[:axi] + (sl_left,) + slice_3d[axi+1:]
         slice_3dr = slice_3d[:axi] + (sl_right,) + slice_3d[axi+1:]
         def func(field, data):
-            ds = div_fac * data[ftype, "d%s" % ax]
-            f  = data[grad_field][slice_3dr]/ds[slice_3d]
-            f -= data[grad_field][slice_3dl]/ds[slice_3d]
+            reorder = getattr(data, '_gradient_swap_axes', False)
+            if reorder:
+                field_data = data[grad_field].swapaxes(0, 2)
+            else:
+                field_data = data[grad_field]
+            ds = data[ftype, "d%s" % ax]
+            vl = field_data[slice_3dl]
+            vc = field_data[slice_3d]
+            vr = field_data[slice_3dr]
+
+            okl = np.isfinite(vl)
+            okrl = okl & np.isfinite(vr)
+
+            f = np.where(okrl, (vr - vl) / 2,
+                np.where(okl, vc - vl, vr - vc)) / ds[slice_3d]
+
             new_field = np.zeros_like(data[grad_field], dtype=np.float64)
-            new_field = data.ds.arr(new_field, f.units)
+            new_field = data.ds.arr(new_field, vr.units / ds.units)
```

Comment:
Any chance you'd consider changing `ds` here to something else, that isn't overloaded elsewhere?  I had to check was type `ds` was since we usually use it for `Dataset`.